### PR TITLE
raftstore: block message when applying snapshot

### PR DESCRIPTION
**Pull request by BusyJay**
So leader's progress won't be changed if follower has not applied the snapshot yet. Merge can't be reliable if the progresses in leader can be false positive.

/cc @disksing, Pd may be affected too, because this can make pending state last for a longer time.

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -246,6 +246,10 @@ pub struct Peer {
     //      locally.
     leader_lease_expired_time: Option<Either<Timespec, Timespec>>,
 
+    // If a snapshot is being applied asynchronously, messages should not be sent untill
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: do we need to use `until` here? @queenypingcap

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -726,6 +731,14 @@ impl Peer {
             return;
         }
 
+        if !self.pending_messages.is_empty() {
+            let messages = mem::replace(&mut self.pending_messages, vec![]);
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: should add a fail point here

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -246,6 +246,10 @@ pub struct Peer {
     //      locally.
     leader_lease_expired_time: Option<Either<Timespec, Timespec>>,
 
+    // If a snapshot is being applied asynchronously, messages should not be sent untill
+    // application is finished.
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: application? maybe the snapshot applying?

**Comment by siddontang** [comment]
any test to cover it?

File: src/raftstore/store/peer.rs (hunk 1 above)
Comment (queenypingcap) [reply src/raftstore/store/peer.rs]: Yes. But please use "until" instead of "untill".

**Comment by BusyJay** [comment]
PTAL

File: tests/failpoints.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+use std::sync::*;
+
+lazy_static! {
+    /// Failpoints are global structs, hence rules set in different cases
+    /// may affect each other. So use a global lock to synchronize them.
+    static ref LOCK: Mutex<()> = Mutex::new(());
+}
+
+fn setup<'a>() -> MutexGuard<'a, ()> {
```
Comment (siddontang) [inline tests/failpoints.rs]: so we can only run the fail point case one by one only?

File: Makefile [hunk 5]
```diff
@@ -30,7 +30,7 @@ default: release
 all: format build test
 
 dev:
-	@export ENABLE_FEATURES=dev && make all
+	@env ENABLE_FEATURES=dev FAIL_POINT=1 make all
```
Comment (siddontang) [inline Makefile]: do we need adding fail point in CI too?

File: tests/failpoints.rs (hunk 4 above)
Comment (BusyJay) [reply tests/failpoints.rs]: Yes.

**Comment by siddontang** [comment]
PTAL @overvenus

**Comment by siddontang** [comment]
LGTM

**Comment by BusyJay** [comment]
ping

**Comment by siddontang** [comment]
please fix the conflict @BusyJay

**Pull request merged**